### PR TITLE
Change bcrypt library to at.favre.lib/bcrypt

### PR DESCRIPTION
**Pull request by johannesloetzsch**
Now supporting salt versions 2b,2x,2y

https://github.com/patrickfav/bcrypt#enhancements-over-jbcrypt

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -1,22 +1,24 @@
-(defproject crypto-password "0.2.0"
+(defproject crypto-password "0.2.1"
```
Comment (weavejester) [inline project.clj]: Version number change needs to be removed.

File: project.clj [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
   :description "Library for securely hashing passwords"
   :url "https://github.com/weavejester/crypto-password"
   :license {:name "Eclipse Public License"
             :url "http://www.eclipse.org/legal/epl-v10.html"}
-  :dependencies [[org.clojure/clojure "1.5.1"]
+  :dependencies [[org.clojure/clojure "1.10.0"]
```
Comment (weavejester) [inline project.clj]: No need to change the minimum Clojure version, as far as I can see?

File: project.clj [hunk 3]
```diff
@@ -1,22 +1,24 @@
...
+                 [at.favre.lib/bcrypt "0.7.0"]
                  [com.lambdaworks/scrypt "1.4.0"]]
   :plugins [[lein-codox "0.9.4"]]
   :codox
   {:output-path "codox"
    :project  {:name "Crypto-Password"}
    :metadata {:doc/format :markdown}
    :source-uri "http://github.com/weavejester/crypto-password/blob/{version}/{filepath}#L{line}"}
-  :aliases {"test-all" ["with-profile" "+1.6:+1.7:+1.8" "test"]}
+  :aliases {"test-all" ["with-profile" "+1.6:+1.7:+1.8:+1.9:+1.10" "test"]}
```
Comment (weavejester) [inline project.clj]: Adding more test profiles is a good change, but needs to be in a separate commit and PR.

File: src/crypto/password/bcrypt.clj [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
-  See: http://bcrypt.sourceforge.net/"
-  (:import org.mindrot.jbcrypt.BCrypt))
+  See: https://github.com/patrickfav/bcrypt"
+  (:import [at.favre.lib.crypto.bcrypt BCrypt]))
 
 (defn encrypt
   "Encrypt a password string using the BCrypt algorithm. The optional work
   factor is the log2 of the number of hashing rounds to apply. The default
-  work factor is 11."
+  work factor is 12."
```
Comment (weavejester) [inline src/crypto/password/bcrypt.clj]: What's the reason behind this change?

**Comment by johannesloetzsch** [comment]
Thank you for your fast response. I appreciate your feedback :)
I force pushed separate commits considering your hints.

File: project.clj (hunk 2 above)
Comment (johannesloetzsch) [reply project.clj]: you are right

File: src/crypto/password/bcrypt.clj (hunk 4 above)
Comment (johannesloetzsch) [reply src/crypto/password/bcrypt.clj]: I tried considering the recommendations by OWASP https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Password_Storage_Cheat_Sheet.md#work-factor having in mind this benchmarks: https://wildlyinaccurate.com/bcrypt-choosing-a-work-factor/

12 Rounds is the default by BSDs PassLib https://passlib.readthedocs.io/en/stable/lib/passlib.hash.bcrypt.html

In case of doubt I vote for a conservative default

Comment (weavejester) [reply src/crypto/password/bcrypt.clj]: As mentioned in the README, the work factor for crypto-password was chosen by benchmarking on AWS machines. I don't mind changing it, but it should be a separate PR that updates all of the work factors, based on a new set of benchmarks.

File: project.clj [hunk 5]
```diff
@@ -6,17 +6,19 @@
   :dependencies [[org.clojure/clojure "1.5.1"]
                  [crypto-random "1.2.0"]
                  [crypto-equality "1.0.0"]
-                 [commons-codec "1.10"]
-                 [org.mindrot/jbcrypt "0.3m"]
+                 [commons-codec "1.12"]
```
Comment (weavejester) [inline project.clj]: Why has this been updated?

File: src/crypto/password/bcrypt.clj (hunk 4 above)
Comment (johannesloetzsch) [reply src/crypto/password/bcrypt.clj]: ok

File: project.clj (hunk 5 above)
Comment (johannesloetzsch) [reply project.clj]: Just to be up to date instead of using the 5year old version.
I now removed it from this PR to keep the changes separated.

Comment (weavejester) [reply project.clj]: It's a good idea, but PRs should be focused on a single change.
